pcsx2-qt/VCRuntimeChecker.cpp

```diff
@@ -0,0 +1,108 @@
+// SPDX-FileCopyrightText: 2002-2023 PCSX2 Dev Team
+// SPDX-License-Identifier: LGPL-3.0+
+
+#include "common/RedtapeWindows.h"
+#include <shellapi.h>
+
+#include "fmt/format.h"
+
+// Minimum version is 14.38.33135.0.
+static constexpr u32 MIN_VERSION_V0 = 14;
+static constexpr u32 MIN_VERSION_V1 = 38;
+static constexpr u32 MIN_VERSION_V2 = 33135;
+static constexpr u32 MIN_VERSION_V3 = 0;
+static constexpr const char* DOWNLOAD_URL = "https://aka.ms/vs/17/release/vc_redist.x64.exe";
+
+struct VCRuntimeCheckObject
+{
+	VCRuntimeCheckObject()
+	{
+		const HMODULE crt_handle = GetModuleHandleW(L"msvcp140.dll");
+		if (!crt_handle)
+			return;
+
+		const HANDLE heap = GetProcessHeap();
+		DWORD filename_length = MAX_PATH;
+		LPWSTR filename = static_cast<LPWSTR>(HeapAlloc(heap, 0, filename_length));
+		if (!filename)
+			return;
+
+		for (;;)
+		{
+			DWORD len = GetModuleFileNameW(crt_handle, filename, filename_length);
+			if (len == filename_length && GetLastError() == ERROR_INSUFFICIENT_BUFFER)
+			{
+				filename_length *= 2;
+				if (filename_length >= 4 * 1024)
+					return;
+				LPWSTR new_filename = static_cast<LPWSTR>(HeapReAlloc(heap, 0, filename, filename_length));
+				if (!new_filename)
+				{
+					HeapFree(heap, 0, filename);
+					return;
+				}
+				filename = new_filename;
+				continue;
+			}
+
+			break;
+		}
+
+		const DWORD version_size = GetFileVersionInfoSizeExW(0, filename, nullptr);
+		LPVOID version_block;
+		if (version_size == 0 || !(version_block = HeapAlloc(heap, 0, version_size)))
+		{
+			HeapFree(heap, 0, filename);
+			return;
+		}
+
+		VS_FIXEDFILEINFO* fi;
+		UINT fi_size;
+		if (!GetFileVersionInfoExW(0, filename, 0, version_size, version_block) ||
+			!VerQueryValueW(version_block, L"\\", reinterpret_cast<LPVOID*>(&fi), &fi_size))
+		{
+			HeapFree(heap, 0, version_block);
+			HeapFree(heap, 0, filename);
+			return;
+		}
+
+		const DWORD v0 = (fi->dwFileVersionMS >> 16) & 0xFFFFu;
+		const DWORD v1 = fi->dwFileVersionMS & 0xFFFFu;
+		const DWORD v2 = (fi->dwFileVersionLS >> 16) & 0xFFFFu;
+		const DWORD v3 = fi->dwFileVersionLS & 0xFFFFu;
+
+		HeapFree(heap, 0, version_block);
+		HeapFree(heap, 0, filename);
+
+		if (v0 >= MIN_VERSION_V0 && v1 >= MIN_VERSION_V1 && v2 >= MIN_VERSION_V2)
```

Comment:
This can cause some issues in the future. If V0 is strictly greater than MIN_VERSION_V0, there is no reason to check for V1 and V2 (and same for V1 with V2).
For example msvc version 14.39.0.0 would fail this check and display the warning prompt, which I don't think is intended.

Reply:
I think this is a common trip up when doing version number checks, you need to do

v0 > Min_version0 || (v0 == min_version0 && v1 >= min_version1 && v2 >= min_version2)